lib/manager/regex/index.ts

```diff
@@ -1,55 +1,161 @@
+  }
+  dependency.replaceString = String(matchResult[0]);
+  return dependency;
+}
+
+function mergeDependency(deps: PackageDependency[]): PackageDependency {
+  const result: PackageDependency = {};
+  deps.forEach((dep) => {
+    fields.forEach((field) => {
+      if (dep[field]) {
+        result[field] = dep[field];
+        // save the line replaceString of the section which contains the current Value for a speed up lookup during the replace phase
+        if (field === 'currentValue') {
+          result.replaceString = dep.replaceString;
         }
       }
-      dep.replaceString = String(matchResult[0]);
-      deps.push(dep);
+    });
+  });
+  return result;
+}
+
+function handleAny(
+  content: string,
+  packageFile: string,
+  config: CustomExtractConfig
+): PackageDependency[] {
+  return config.matchStrings
+    .map((matchString) => regEx(matchString, 'g'))
+    .flatMap((regex) => regexMatchAll(regex, content)) // match all regex to content, get all matches, reduce to single array
+    .map((matchResult) => createDependency(matchResult, config));
+}
+
+function handleCombination(
+  content: string,
+  packageFile: string,
+  config: CustomExtractConfig
+): PackageDependency[] {
+  const dep = handleAny(content, packageFile, config).reduce(
+    (mergedDep, currentDep) => mergeDependency([mergedDep, currentDep]),
+    {}
+  ); // merge fields of dependencies
+  return [dep];
+}
+
+function handleRecursive(
+  content: string,
+  packageFile: string,
+  config: CustomExtractConfig,
+  index = 0
+): PackageDependency[] {
+  const regexes = config.matchStrings.map((matchString) =>
+    regEx(matchString, 'g')
+  );
+  // abort if we have no matchString anymore
+  if (regexes[index] == null) {
+    return [];
+  }
+  return regexMatchAll(regexes[index], content).flatMap((match) => {
+    // if we have a depName and a currentValue with have the minimal viable definition
+    if (match?.groups?.depName && match?.groups?.currentValue) {
+      return createDependency(match, config);
     }
-  } while (matchResult);
-  if (deps.length) {
-    return { deps, matchStrings: config.matchStrings };
+    return handleRecursive(match[0], packageFile, config, index + 1);
+  });
+}
+
+export function extractPackageFile(
+  content: string,
+  packageFile: string,
+  config: CustomExtractConfig
+): Result<PackageFile | null> {
+  let deps;
+  switch (config.matchStringsStrategy) {
+    default:
+    case 'any':
+      deps = handleAny(content, packageFile, config);
+      break;
+    case 'combination':
+      deps = handleCombination(content, packageFile, config);
+      break;
+    case 'recursive':
+      deps = handleRecursive(content, packageFile, config);
+      break;
   }
+
+  deps = deps.filter((dep) => {
+    return dep != null;
+  });
+  if (deps?.length) {
+    if (config.matchStringsStrategy) {
+      return {
+        deps,
+        matchStrings: config.matchStrings,
+        matchStringsStrategy: config.matchStringsStrategy,
```

Comment:
Is this necessary later on? I thought it should be inherited from `config` anyway

Reply:
Renovate seems to change the branch and then run the manager again to confirm the change applied is the same as the result from the manager.     
But in this second go trough `matchStringsStrategy` is undefined which leads to that an `any` strategy is assumed even if the value has been `recursive` or `combination`.

This additional return value fixes this.